server/src/main/java/org/apache/druid/server/coordinator/duty/KillDatasourceMetadata.java

```diff
@@ -0,0 +1,110 @@
+ * under the License.
+ */
+
+package org.apache.druid.server.coordinator.duty;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+import com.google.inject.Inject;
+import org.apache.druid.indexing.overlord.IndexerMetadataStorageCoordinator;
+import org.apache.druid.indexing.overlord.supervisor.NoopSupervisorSpec;
+import org.apache.druid.indexing.overlord.supervisor.SupervisorSpec;
+import org.apache.druid.java.util.common.logger.Logger;
+import org.apache.druid.java.util.emitter.service.ServiceEmitter;
+import org.apache.druid.java.util.emitter.service.ServiceMetricEvent;
+import org.apache.druid.metadata.MetadataSupervisorManager;
+import org.apache.druid.server.coordinator.DruidCoordinatorConfig;
+import org.apache.druid.server.coordinator.DruidCoordinatorRuntimeParams;
+
+import java.util.Collection;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * CoordinatorDuty for automatic deletion of datasource metadata from the datasource table in metadata storage.
+ * (Note: datasource metadata only exists for datasource created from supervisor).
+ */
+public class KillDatasourceMetadata implements CoordinatorDuty
+{
+  private static final Logger log = new Logger(KillDatasourceMetadata.class);
+
+  private final long period;
+  private final long retainDuration;
+  private long lastKillTime = 0;
+
+  private final IndexerMetadataStorageCoordinator indexerMetadataStorageCoordinator;
+  private final MetadataSupervisorManager metadataSupervisorManager;
+
+  @Inject
+  public KillDatasourceMetadata(
+      DruidCoordinatorConfig config,
+      IndexerMetadataStorageCoordinator indexerMetadataStorageCoordinator,
+      MetadataSupervisorManager metadataSupervisorManager
+  )
+  {
+    this.indexerMetadataStorageCoordinator = indexerMetadataStorageCoordinator;
+    this.metadataSupervisorManager = metadataSupervisorManager;
+    this.period = config.getCoordinatorDatasourceKillPeriod().getMillis();
+    Preconditions.checkArgument(
+        this.period >= config.getCoordinatorMetadataStoreManagementPeriod().getMillis(),
+        "Coordinator datasource metadata kill period must be >= druid.coordinator.period.metadataStoreManagementPeriod"
+    );
+    this.retainDuration = config.getCoordinatorDatasourceKillDurationToRetain().getMillis();
+    Preconditions.checkArgument(this.retainDuration >= 0, "Coordinator datasource metadata kill retainDuration must be >= 0");
+    log.debug(
+        "Datasource Metadata Kill Task scheduling enabled with period [%s], retainDuration [%s]",
+        this.period,
+        this.retainDuration
+    );
+  }
+
+  @Override
+  public DruidCoordinatorRuntimeParams run(DruidCoordinatorRuntimeParams params)
+  {
+    if ((lastKillTime + period) < System.currentTimeMillis()) {
+      lastKillTime = System.currentTimeMillis();
+      long timestamp = System.currentTimeMillis() - retainDuration;
```

Comment:
nit: use consistent timestamp for all calculations

```suggestion
    long currentTimeMillis = System.currentTimeMillis();
    if ((lastKillTime + period) < currentTimeMillis) {
      lastKillTime = currentTimeMillis;
      long timestamp = currentTimeMillis - retainDuration;
```

Reply:
Additional question - I notice this pattern in a few other co-ordinator duties.

Are there any additional safeguards we need for a very large `retainDuration`? What happens if `timestamp` is calculated to be negative?

Reply:
Added safeguard so that we never get calculated `timestamp` to be negative

Reply:
Note that I intentionally didnt specific in the docs that retainDuration have to be less than current timestamp (although we do check against this condition to protect ourself from unexpected behavior), since this is a very unlikely scenario and I don't want to make the docs to be unnecessary verbose.